components/Init/src/virtio_con.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2022, UNSW (ABN 57 195 873 179)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <autoconf.h>
+
+#include <utils/util.h>
+
+#include <camkes.h>
+#include <camkes/dataport.h>
+
+#include <sel4vmmplatsupport/drivers/virtio_con.h>
+#include <sel4vmmplatsupport/device.h>
+
+#include <sel4vm/guest_irq_controller.h>
+#include <sel4vm/boot.h>
+
+#include <platsupport/serial.h>
+
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
+
+#include "virtio_con.h"
+#include "virtio_irq.h"
+
+static virtio_con_t *virtio_con;
+
+typedef struct virtio_con_cookie {
+    virtio_con_t *virtio_con;
+    vm_t *vm;
+} virtio_con_cookie_t;
+
+/* Configured in camkes */
+#define NUM_PORTS ARRAY_SIZE(serial_layout)
+
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE 4088
```

Comment:
Is it 4096 bytes or one page? If it's one page, then use the PAGESIZE macro (whatever it's called)

Reply:
It's 4096 bytes. CAmkES demands shmem buffers be a multiple of 0x1000 bytes.